### PR TITLE
fix: Recursive search for versions of shared dependencies

### PR DESCRIPTION
**Pull request by ScriptedAlchemy**
Improved recursive version search of shared dependencies for cases where there is a parent `package.json` containing the required version. This is specific to monorepos and dual packages(`esm`/`cjs`), where the closest `package.json` is only used to specify the package type.  

https://github.com/webpack/webpack/issues/13457

**What kind of change does this PR introduce?**

bugfix
**Did you add tests for your changes?**

yes
**Does this PR introduce a breaking change?**

no
**What needs to be documented once your changes are merged?**
none

### CODE REVIEW

**Review by alexander-akait — APPROVED** [review]

**Comment by ScriptedAlchemy** [comment]
There is an update in want to make to this.

**Comment by ScriptedAlchemy** [comment]
Okay update pushed

File: lib/sharing/utils.js [hunk 1]
```diff
@@ -320,25 +320,48 @@ function normalizeVersion(versionDesc) {
 
 module.exports.normalizeVersion = normalizeVersion;
 
-/** @typedef {{ data: JsonObject, path: string }} DescriptionFile */
+/** @typedef {{ data: Record<string, any>, path: string }} DescriptionFile */
```
Comment (alexander-akait) [inline lib/sharing/utils.js:323]: @ScriptedAlchemy Why we changed it?

Comment (ScriptedAlchemy) [reply lib/sharing/utils.js:323]: My bad will correct

**Review by alexander-akait — APPROVED** [review]

**Review by alexander-akait — APPROVED** [review]

**Comment by alexander-akait** [comment]
Thank you

**Pull request merged**